### PR TITLE
initial contribution nuspell

### PR DESCRIPTION
**Pull request by PanderMusubi**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----
closes https://github.com/Homebrew/linuxbrew-core/issues/19713

### CODE REVIEW

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: These are the Catch2 tests, if these fail, installation is not useful. Is that OK?

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: Other systems have packages which supply these files. This was copied from Hunspell and the URL was updated. For example for Debian/Ubuntu, the aff and dic files are available in other packages which can be installed and used for further testing. Is that OK?

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: For that a .dic and .aff file is needed and those are not available in homebrew. And we don't want to use a wget/curl to get them into the build process. Would `nuspell -D` be all right here as found in the latest commit? It does much more than a `--help` or `--version`. See also the answer to your next question below.

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: Homebrew doesn't run testsuites as part of the install but has it's own functional tests at the bottom of the formula.

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: if you get a dictionary as a resource you can use it in the test.

**Review by PanderMusubi — COMMENTED** [review]
moved answer above

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: What kind of directory do you mean? We can't ship the dic and aff file in our code and pulling in stuff from the Internet is not allowed in these build processes.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: The Catch2 test in the building/installing part is part of the build process. It uses unit tests. If these fail, it is of no use to install it and test it later. It is not a test suite that calls the parts that have been build, but directly the code inside, before the build is complete. It is an internal build test. OK to leave it there?

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: The Catch2 test in the building/installing part is part of the build process. It uses unit tests. If these fail, it is of no use to install it and test it later. It is not a test suite that calls the parts that have been build, but directly the code inside, before the build is complete. It is an internal build test. OK to leave it there?

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: We could generate a very minimalistic on the fly, but I need some help with that putting it in the Formula.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: Removed

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: Nuspell needs C++-17, building CI linuxbrew-core repo needed this. Nuspell works with gcc or clang. On my machine I don't need to mention it either. I have removed it.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: Nuspell needs C++-17, building CI linuxbrew-core repo needed this. Nuspell works with gcc or clang. On my machine I don't need to mention it either. I have removed it.

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: Pulling stuff from the internet for a test isn't allowed by upstream?

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: That's what the homebrew test block is for. We don't test if software passes their test suite, we test if it works as expected for the users.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: So moving `system "cmake", "--build", ".", "--target", "test"` to the test block would suffice? It (also) does run a lot of tests like a user would run them.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: It is not in package building downstream, so you can build air-gapped and prevent injection from third parties.

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: That won't work since the test doesn't have access to the source, it's better to just remove this.

File: Formula/nuspell.rb
Comment (SMillerDev) [reply Formula/nuspell.rb]: I'm not sure what you mean by that but homebrew downloads tons of stuff just for test runs https://github.com/Homebrew/homebrew-core/blob/master/Formula/hapi-fhir-cli.rb#L24-L28 is an example

Comment (PanderMusubi) [reply Formula/nuspell.rb]: Okay, if that is allowed, I can add a realistic user test.

**Comment by SMillerDev** [comment]
doesn't install on mojave

**Comment by PanderMusubi** [comment]
@SMillerDev please review latest commit. I've added a test as requested.

Is this also the proper way to exclude building for mojave (and still allowing for building in Linuxbrew)?

File: Formula/nuspell.rb
Comment (SMillerDev) [inline Formula/nuspell.rb]: this is only needed for `:test`

File: Formula/nuspell.rb
Comment (SMillerDev) [inline Formula/nuspell.rb]: if it doesn't build on all supported macOS versions it should be in a tap and not in homebrew-core.

Comment (PanderMusubi) [reply Formula/nuspell.rb]: Ah, you wrote me not to build for mojave, at least, that is how I read that. The problem for mojave is in the ruby dependency, not nuspell. I will remove that line so the error pops up in the CI.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: Thanks. Meanwhile, have switched to binutils as that is more common to be installed.

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: As you can see, 10.14 fails on https://github.com/Homebrew/homebrew-core/pull/50685/checks?check_run_id=470015012#step:4:432 Do you know how that can be fixed?

Comment (PanderMusubi) [reply Formula/nuspell.rb]: Can you also see why continuous-integration/jenkins/ghprb has failed? Looks like an issue with GitHub.

File: Formula/firebase-cli.rb
Comment (SMillerDev) [inline Formula/firebase-cli.rb]: Please don't remove other formula

File: Formula/nuspell.rb
Comment (SMillerDev) [inline Formula/nuspell.rb]: Is this just counting words?

**Review by SMillerDev — CHANGES_REQUESTED** [review]

File: Formula/nuspell.rb
Comment (Bo98) [reply Formula/nuspell.rb]: For 10.14, you might need either `ENV.delete("SDKROOT")` or [something like this](https://github.com/Homebrew/homebrew-core/blob/0745f27ed2eee8a8f62248938a8b8acdbd6002c2/Formula/weechat.rb#L37-L39). It's something I need to look into more at some point.

**Comment by SMillerDev** [comment]
Now that I actually look at the audits instead of the file I notice:

> GitHub repository not notable enough (<30 forks, <30 watchers and <75 stars)

And while this seems like a nice software, the above audit means it's not notable enough for us to package in Homebrew for now. This is not a decision on the merits of the software, but we cannot distribute every single piece of software and have to draw a line on “notability” somewhere. Of course, we would be happy to reconsider in the future.

Please consider hosting it in a personal tap, which is very easy to do: https://docs.brew.sh/How-to-Create-and-Maintain-a-Tap

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: It counts how many words are spelled correctly of the 50,000 words that are checked. In that way the test will still work if we improve or change the spell checking. The margin can be much less, but this is a safe setting and is a real-world example.

File: Formula/firebase-cli.rb
Comment (PanderMusubi) [reply Formula/firebase-cli.rb]: Is was not aware of the that change, not sure where originated. was not intended at all.

**Comment by PanderMusubi** [comment]
@SMillerDev Please reconsider this PR for the following reasons:
* have a look at the notability of https://github.com/hunspell/hunspell/ as Nuspell is the drop-in replacement for Hunspell.
* Nuspell will be used by Firefox, Enchant, LibreOffice and much much more. We got two large grants of Mozilla Open Source Support in order to do so, see https://nuspell.github.io/
* Enchant is releasing a new version soon with Nuspell integration, see https://github.com/AbiWord/enchant/pull/228 Enchant is also in Homebrew and has also a high enough Notability.
* Many maintainers of major (and less popular) distributions are including Nuspell as you can see here https://github.com/nuspell/nuspell/wiki/Nuspell-packaged-binaries

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: @Bo98 Thanks for the advice. Is that only for mojave `==` or for mojave and above `>=`?

Comment (Bo98) [reply Formula/nuspell.rb]: Either. It currently makes no difference on Catalina - it will do when Xcode 12 is released. But I hope to have fixed it properly by then so these workarounds aren't needed.

**Comment by SMillerDev** [comment]
While I appreciate the effort, the [self submission](https://docs.brew.sh/Acceptable-Formulae#niche-or-self-submitted-stuff) in combination with the lacking notability make me think it's not quite there yet. Though if you want I'd be happy to help you get a formula in a tap to a state where all you have to do is migrate it once you meet the threshold.

**Comment by PanderMusubi** [comment]
Okay, when the notability is sufficient, I will rerequest this. You help so far is very much appreciated. I'll follow the documentation on creating a tap and check back with you once it is up .

File: Formula/nuspell.rb
Comment (PanderMusubi) [reply Formula/nuspell.rb]: Thanks, do you have an issue on that which I can follow?

**Comment by PanderMusubi** [comment]
@SMillerDev is this done properly? https://github.com/nuspell/homebrew-nuspell

**Comment by SMillerDev** [comment]
Looks good, though for eventual homebrew core inclusion I'd like to see the test block do spell checking instead of just reading the word lists.

**Comment by PanderMusubi** [comment]
Thanks. Well, it actually does spell checking. It loads a library and checks a list of 50.000 words. If these words are in a list or in a running text doesn't matter for the spell checking. What else would you like to see?

**Pull request closed without merging**
